src/HotChocolate/Core/src/Execution/Batching/BatchExecutor.Enumerable.cs

```diff
@@ -46,9 +50,80 @@ private class BatchExecutorEnumerable : IAsyncEnumerable<IQueryResult>
             _inputFormatter = inputFormatter ??
                 throw new ArgumentNullException(nameof(inputFormatter));
             _visitor = new CollectVariablesVisitor(requestExecutor.Schema);
+            _allowParallelExecution = allowParallelExecution;
         }
 
-        public async IAsyncEnumerator<IQueryResult> GetAsyncEnumerator(
+        public IAsyncEnumerator<IQueryResult> GetAsyncEnumerator(
+            CancellationToken cancellationToken = default)
+        {
+            if (_allowParallelExecution)
+            {
+                return GetParallelAsyncEnumerator(cancellationToken);
+            }
+            else
+            {
+                return GetSequentialAsyncEnumerator(cancellationToken);
+            }
+        }
+
+        private async IAsyncEnumerator<IQueryResult> GetParallelAsyncEnumerator(
+            CancellationToken cancellationToken = default)
+        {
+            var pending = new List<Task<IQueryResult>>();
+            var exportCount = _exportedVariables.Count;
+            var hasMutatedState = false;
+            BatchScheduler? scheduler = null;
+            foreach (IQueryRequest queryRequest in _requestBatch)
+            {
+                var nextScheduler = queryRequest.Services.GetRequiredService<IBatchDispatcher>() as BatchScheduler;
+                if (scheduler != nextScheduler)
+                {
+                    if (scheduler != null)
+                    {
+                        throw new InvalidOperationException("Too many schedulers");
+                    }
+
+                    scheduler = nextScheduler;
+                }
+                scheduler?.Suspend();
+                var request = (IReadOnlyQueryRequest)queryRequest;
+                var canMutateState = CanAffectState(request);
+                if (_exportedVariables.Count > exportCount || canMutateState || hasMutatedState)
```

Comment:
This logic tries to detect that either the operation can mutate state. So if there is a mutation, we will run everything in parallel up to that point, then run mutations in sequence, then we can run queries in parallel after that. If a variable was exported by the previous batch, we await before running the next batch(es).